### PR TITLE
Porting go-ethereum/rlp package.

### PR DESCRIPTION
**Pull request by sirano11**
## Proposed changes

For using "optional" struct tag, ported latest go-ethereum/rlp package.
  - Among the packages imported and referenced by the rlp package, Changed the ethereum package that referenced from rlp to the klaytn package.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by kjhman21** [comment]
@sirano11 Could you explain what you've changed in addition to bringing the original Ethereum code?

**Comment by sirano11** [comment]
> Could you explain what you've changed in addition to bringing the original Ethereum code?
@kjhman21 
Among the packages imported and referenced by the rlp package, I changed the ethereum package to the klaytn package.

**Review by kjhman21 — APPROVED** [review]
LGTM. 
@aidan-kwon Please take a look.

**Review by kjeom — DISMISSED** [review]

**Comment by kjeom** [comment]
I tested encoding/decoding `block header` before/after `KIP-71 hard fork` and It works well :)

File: rlp/internal/rlpstruct/rlpstruct.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+}
+
+// ProcessFields filters the given struct fields, returning only fields
+// that should be considered for encoding/decoding.
+func ProcessFields(allFields []Field) ([]Field, []Tags, error) {
+	lastPublic := lastPublicField(allFields)
+
+	// Gather all exported fields and their tags.
+	var fields []Field
+	var tags []Tags
```
Comment (hqjang-pepper) [inline rlp/internal/rlpstruct/rlpstruct.go:109]:
```suggestion
	fields := make([]Field, 0)
	tags := make([]Tags, 0)
```
suggestion to fix prealloc problem from test-linter

Comment (sirano11) [reply rlp/internal/rlpstruct/rlpstruct.go:109]: Thanks. @hqjang-pepper.

**Comment by hyunsooda** [comment]
Based on this big change, could you describe how you tested compatibility tests? For example:
- Given a stored block (or any structure) with the previous RLP encoding method, is it decodable with the new RLP decoder?
- By the new changes (`optional fields`, `nil`, `nilList`, `nilString`, etc), Is there a case that the RLP-encoded data is misinterpreted between two nodes whose have different RLP encoding and decoding methods) (https://github.com/klaytn/klaytn/pull/1385/files#diff-0307e7f9d52cb5d1be9e58e3ed38f13060d2b28e85617df74aed30b08a720445L239-R353)
- etc

As you know better than I, the RLP is widely used in the existing functionalities (e.g., network, DB, etc.). The facts seem to be required that this new change perfectly works in all the places used if the fundamental rules of RLP are still preserved (The rules seem preserved in my view) .

**Comment by hyunsooda** [comment]
Could you take a look again entirely the imported license? (some of the files seems not updated and even not imported)

**Comment by sirano11** [comment]
@hyunsooda 

> Given a stored block (or any structure) with the previous RLP encoding method, is it decodable with the new RLP decoder?

Because it already exists a test about rlp encoding / decoding, i think it is enough to check compatiblility with testing existed test case.

> By the new changes (optional fields, nil, nilList, nilString, etc), Is there a case that the RLP-encoded data is misinterpreted between two nodes whose have different RLP encoding and decoding methods?) (https://github.com/klaytn/klaytn/pull/1385/files#diff-0307e7f9d52cb5d1be9e58e3ed38f13060d2b28e85617df74aed30b08a720445L239-R353)

These changes will be contained hardfork update, it means that all node must update new klaytn version which contains these changes.


> Could you take a look again entirely the imported license? (some of the files seems not updated and even not imported)

Only files that I have modified (eg, change of import path) have modified the license, and files that have not been modified do not change the license part.
If I made a mistake, can you tell me in the comments?

**Comment by hyunsooda** [comment]
Okay. Then please describe this change is for the hardfork update to not be popped this similar questions.  

> Because it already exists a test about rlp encoding / decoding, i think it is enough to check compatiblility with testing existed test case.

All the current test units call the newly replaced RLP encoding/decoding methods. I was concerned the surprising difference in encoding and decoding between the old one and your change may lead to wrong encoding/decoding.

> These changes will be contained hardfork update, it means that all node must update new klaytn version which contains these changes.

Even if its change is for the hardfork, EN/PN binaries must be updated since its change is not bounded to consensus/governance. 

> Only files that I have modified (eg, change of import path) have modified the license, and files that have not been modified do not change the license part.
If I made a mistake, can you tell me in the comments?

- If you imported `testdata` directory as is, putting licenses are required, or if it is generated by tools or code, no change seems required.
- `types.go`, `gen.go`, `gen_test.go`, `enc_buffer.go`, `safe.go`, `unsafe.go`, 

Hope to hear your comment again.

**Review by kjhman21 — APPROVED** [review]
LGTM

**Review by kjeom — APPROVED** [review]

**Review by hqjang-pepper — APPROVED** [review]

**Comment by hyunsooda** [comment]
@sirano11, Could you post an issue and address the unexpected cases such as https://github.com/klaytn/klaytn/pull/1385/files#diff-0307e7f9d52cb5d1be9e58e3ed38f13060d2b28e85617df74aed30b08a720445L239-R353?

**Pull request merged**